### ISSUE TITLE
[FEA] have cudf::merge or similar API be stable, with the order of the tables passed in

### ISSUE THREAD

**Issue by revans2** [rapidsai/cudf#7379]
**Is your feature request related to a problem? Please describe.**
Spark's sort is a stable sort.  Most of the time that does not mater much, but there are a few cases where it does make a difference in the output (some window functions in a single process).

`cudf::merge` currently provides no way to ensure that if batch 1 came before batch 2 in read order that the output would be preserved.

**Describe the solution you'd like**
It would be great in the case of ties in cudf::merge if the order of the rows could match the order of the tables passed in. If this is going to cause a performance difference with the existing merge then a separate API is fine too.

**Describe alternatives you've considered**
Add in a separate row of longs with all of the data that is a sequence to get a total ordering of the data read in so I can then use it as the final key in the sort, but that adds a lot of extra memory for something that is a really rare use case.

**Additional context**
This is not super critical, as I said it is for a few corner cases. But from looking at the code it looks like it is simple enough that I could do it. The underlying `trust::merge` says that it is stable for what we want in the docs. That just leaves the order in which the tables are merged. It looks like right now there is a priority queue used to try and reduce the amount of data transferred and merges done by merging the two smallest tables at a time. I think if we instead merged them in waves, we could still reduce the amount of intermediate data transferred and maintain stable ordering.  Not as clean as the existing code but the following pseudo code would probably do what we want.

```
queue input(all the input tables)
queue output()
while (input.size > 1) {
  while (input.size > 1) {
    left = input.pop
    right = input.pop
    output.add(merge(left, right))
  }
  if (input.size == 1) {
    output.add(input.pop)
  }
  input = output
  output.reset()
}
return input.pop
```

**Comment by jrhemstad** [comment]
I don't quite follow the psuedocode. Using a priority queue will break the original input ordering. It looks like you're just suggesting merging `(0,1)` then merginging that result with `2`, e.g., `(((0,1), 2), 3)),...`? Is that right?

That would work, but would defeat the optimization of always merging the smallest tables first (which makes a significant difference).

It should be possible to still maintain the original order and merge the smallest tables first, it will just be a bit complicated. E.g., if we have 5 tables A, B, C, D, E whose sizes are (100, 10, 50, 10, 100) then we could merge `B' = (B,D)`  (size is 20) then `B'' = (B', C)` (size is 70), then `A' = (A, B'')` (size is 170) and finally `Result = (A', E)` (size is 270).

So long as we always merge right to left, that should still result in the same order as if we did `((((A,B), C), D), E)`.  There's probably a clever algorithm to do this nicely. Maybe @codereport can think of it.

**Comment by jlowe** [comment]
> It should be possible to still maintain the original order and merge the smallest tables first

I don't think that sort remains stable.  I'm assuming the final table ordering between tied sort keys needs to be A->E to match the order in which the tables were generated. No matter whether we merge (B,D) with C or C with (B,D), the stable ordering between B, C, and D has been violated.  (B,D) will be treated as a new input table, and there's no way to properly insert the C keys since we lost the ability to distinguish B from D once they merged.  I think we'd have to manifest a discriminator column (e.g.: table ID) to be able to skip table ordering when merging yet maintain a stable sort.

**Comment by jrhemstad** [comment]
> No matter whether we merge (B,D) with C or C with (B,D), the stable ordering between B, C, and D has been violated. (B,D) will be treated as a new input table, and there's no way to properly insert the C keys since we lost the ability to distinguish B from D once they merged.

Yeah, that's a good point. My idea won't work.

> we could still be smart about merging smaller pairs first, but those pairs always have to be adjacent tables in the table ordering.

Yeah, this would work. It may not be as optimal as merging the exact smallest tables first, but it would at least be better than the naive left fold.

**Comment by revans2** [comment]
I do want to reiterate that this is for a minor corner case in Spark. We don't have a customer yet that is complaining about the corner case either. The priority should be fairly low until we do hit a customer complaint.

**Comment by harrism** [comment]
Why not just provide a choice between fast and stable?

**Comment by jlowe** [comment]
This is still desired but not high priority.

**Comment by jrhemstad** [comment]
As @harrism mentioned, I think we can just have `merge` and `stable_merge` APIs to satisfy the different requirements.

**Comment by jlowe** [comment]
Still desired.

**Comment by wence-** [comment]
xref: https://github.com/rapidsai/cudf/issues/16010

**Comment by davidwendt** [comment]
From discussion above, it sounds like a stable merge is possible but will likely be slower.
If that is acceptable, then I think a `cudf::stable_merge` would be doable.

**Comment by vyasr** [comment]
Should we close this or #16010 and just track the work in one place, or is there some difference between the two issues?
